games/xworld/maps/xworld_env.py

```diff
@@ -123,18 +124,73 @@ def set_dims(self, h, w):
         self.entities.append(Entity(type=type, loc=loc, name=name))
         self.changed = True
 
+    def reinstantiate_entity_property(self, entity, property=None, value=None):
+        """
+        Reinstantiate a specified property of an existing entity.
+        If no property is specified, all available properties will be reinstantiated.
+        If no value is provided for a specified property, a valid random value will be
+        used for the specified property and all the rest properties remain unchanged.
+        """
+        def check_or_get_value(valid_value_set):
+            """
+            Check if the given value of the specified property is a valid one, or randomly
+            select one from the valid value set if value is None, and return the value.
+            """
+            if value is None:
+                assert len(valid_value_set) > 0, \
+                    "invalid value set for property %s is provided" % property
+                return random.choice(valid_value_set)
+            else:
+                assert value in valid_value_set, \
+                    "invalid value for property %s is provided" % property
+                return value
+        assert not property or property in entity.__dict__.keys(), \
+            "invalid property name: %s is provided" % property
+
+        if property == "loc":
+            entity.loc = check_or_get_value(self.available_grids)
+            # remove used locations
+            self.available_grids.remove(loc)
+        if not property or property == "name":
+            entity.name = check_or_get_value(self.get_all_possible_names(entity.type))
+            # update id once name is changed
+            entity.id = "%s_%s" % (entity.name, self.entity_nums[entity.type])
+        if not property or property == "asset_path" or property == "name":
+            # update the assert_path once the name is changed
+            entity.asset_path = check_or_get_value(self.items[entity.type][entity.name])
+        if (not property or property == "yaw") \
+                                     and get_flag("visible_radius") and e.type != "block":
+            ## if partially observed, perturb the objects
+            yaw_range = range(-1, 3)
+            entity.yaw = check_or_get_value(yaw_range) * self.PI_2
+        if (not property or property == "scale") \
+                                     and get_flag("visible_radius") and e.type == "goal":
+            scale_range = np.arange(0.5, 1.1, 0.1)
+            entity.scale = check_or_get_value(quantize_value(scale_range, 0.1))
+        if (not property or property == "offset") \
+                                     and get_flag("visible_radius") and e.type == "goal":
+            offset_range = np.arange(0, 1 - (entity.scale if hasattr(entity, 'scale') else 0.4), 0.1)
+            entity.offset = check_or_get_value(quantize_value(offset_range, 0.1))
+        # color is not exposed for reinstantiation
```

Comment:
Why set the 0.1 as the step size for scale and offset?

Reply:
improved